src/Microsoft.AspNet.Mvc.WebApiCompatShim/Conventions/WebApiParameterConventionsApplicationModelConvention.cs

```diff
@@ -17,26 +17,28 @@ public void Apply(ActionModel action)
             {
                 foreach (var parameter in action.Parameters)
                 {
-                    if (parameter.BinderMetadata is IBinderMetadata)
+                    // Some IBindingSourceMetadata attributes like ModelBinder attribute return null 
+                    // as their binding source. Special case to ensure we do not ignore them.
+                    if (parameter.BindingInfo.BindingSource != null || 
+                        parameter.Attributes.OfType<IBindingSourceMetadata>().Any())
                     {
                         // This has a binding behavior configured, just leave it alone.
                     }
                     else if (ValueProviderResult.CanConvertFromString(parameter.ParameterInfo.ParameterType))
                     {
                         // Simple types are by-default from the URI.
-                        parameter.BinderMetadata = new FromUriAttribute();
+                        parameter.BindingInfo.BindingSource = (new FromUriAttribute()).BindingSource;
```

Comment:
THat way there would be two properties which accomplish same thing, one static and one not. I do not want to have the api surface reflect that. Choosing to keep this as is.